common/clock/timer_gate.go

```diff
@@ -0,0 +1,109 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package clock
+
+import (
+	"sync"
+	"time"
+)
+
+type (
+	// TimerGate is a timer interface which fires a timer at a given timestamp.
+	TimerGate interface {
+		// Chan return the channel which will be fired when time is up
+		Chan() <-chan time.Time
+		// FireAfter check will the timer get fired after a certain time
+		FireAfter(t time.Time) bool
+		// Update update the timer gate, return true if update is a success
+		// success means timer is idle or timer is set with a sooner time to fire
+		Update(t time.Time) bool
+		// Close shutdown the timer
+		Stop()
+	}
+
+	timerGateImpl struct {
+		sync.RWMutex
+		timeSource TimeSource
+		timer      Timer
+		fireTime   time.Time
+	}
+)
+
+func NewTimerGate(timeSource TimeSource) TimerGate {
+	t := &timerGateImpl{
+		timer:      timeSource.NewTimer(0),
+		timeSource: timeSource,
+	}
+	// the timer should be stopped when initialized
+	if !t.timer.Stop() {
+		// drain the existing signal if exist
+		// TODO: the drain can be removed after go 1.23
+		<-t.timer.Chan()
```

Comment:
should we change this to avoid blocking?
```
select {
	case <-t.timer.Chan():
	default:
	}
```

Reply:
It's safe to block. As of go 1.23, Stop method always returns `true` if the timer is not read from the channel, so we don't go to this branch. And before go 1.23, this is the correct way to stop a timer and drain the channel.